### ISSUE TITLE
More flexible way to define fallback languages (depending on current language)

### ISSUE THREAD

**Issue by nwinter** [i18next/i18next#748]
We have a lot of translations for `zh-HANS`, some translations for `zh-HANT`, no general `zh` translations (because contributors always just want to focus on the one file they are interested in), and general fallback for `en`. Same thing for `de-DE`, `de-AT`, `de-CH`, and a few other families.

If a translation is not found in `zh-HANT`, I want to use the one for `zh-HANS` before going back to `en`.

Is there a way to do this? Outside of i18next, we've [implemented it here](https://github.com/codecombat/codecombat/blob/master/app/core/utils.coffee#L52-L76):

``` coffeescript
module.exports.i18n = (say, target, language=me.get('preferredLanguage', true), fallback='en') ->
  generalResult = null
  fallBackResult = null
  fallForwardResult = null  # If a general language isn't available, the first specific one will do.
  fallSidewaysResult = null  # If a specific language isn't available, its sibling specific language will do.
  matches = (/\w+/gi).exec(language)
  generalName = matches[0] if matches

  for localeName, locale of say.i18n
    continue if localeName is '-'
    if target of locale
      result = locale[target]
    else continue
    return result if localeName is language
    generalResult = result if localeName is generalName
    fallBackResult = result if localeName is fallback
    fallForwardResult = result if localeName.indexOf(language) is 0 and not fallForwardResult?
    fallSidewaysResult = result if localeName.indexOf(generalName) is 0 and not fallSidewaysResult?

  return generalResult if generalResult?
  return fallForwardResult if fallForwardResult?
  return fallSidewaysResult if fallSidewaysResult?
  return fallBackResult if fallBackResult?
  return say[target] if target of say
  null
```

**Comment by jamuhl** [comment]
The problem we got with using some language-regions 'sideways' on the client we just not have that information:

having language `de-CH` we load `de-CH`, `de`, `en` but have no idea there is a `de-AT`

so not sure there will be such a generic solution of looking up regional packages having same language.

but i have a plan to make the fallback more flexible, pseudo code:

```
fallbackLng: {
  'de-CH': ['fr', 'it', 'en'],
  'zh-HANS': ['zh-HANT', 'en'],
  'zh-HANT': ['zh-HANS', 'en']
  'default' : ['en']
}
```

but not yet really sure this is a good idea...

**Comment by nwinter** [comment]
I like that more flexible fallback option!

**Comment by jamuhl** [comment]
should not be to hard to implement...but not yet sure when i get to it.

**Comment by nwinter** [comment]
Perhaps you can describe the steps we'd need to take to add it and we can see if we can help with a pull request?

**Comment by jamuhl** [comment]
basically you need to allow following object as fallbackLng:
```
fallbackLng: {
  'de-CH': ['fr', 'it', 'en'],
  'zh-HANS': ['zh-HANT', 'en'],
  'zh-HANT': ['zh-HANS', 'en']
  'default' : ['en']
}
```
for: https://github.com/i18next/i18next/blob/master/src/defaults.js#L60
there should be nothing todo

for: https://github.com/i18next/i18next/blob/master/src/LanguageUtils.js#L80
you need to resolve the array if array, the array of the fallback defined object from top having the current code or the languagePart of it or default if neither found

for: https://github.com/i18next/i18next/blob/master/src/Translator.js#L117 
i would resolve the fallbacks in default or if array like today

then i will need to check if no plugin depends of it to be an array...but hopefully all plugins take the toResolveHierarchy - evtl. it would be a good idea to provide a getFallbackLng(code) function

**Comment by jamuhl** [comment]
https://github.com/i18next/i18next/issues/805#issuecomment-251784892

**Comment by jamuhl** [comment]
https://github.com/i18next/i18next/commit/3763ac84d2659e8d112fc9daeccdc869a973a7d5

working on it...

**Comment by jamuhl** [comment]
published in i18next@3.5.0
